web/src/message_list_view.ts

```diff
@@ -37,10 +41,92 @@ import * as sub_store from "./sub_store";
 import * as submessage from "./submessage";
 import {is_same_day} from "./time_zone_util";
 import * as timerender from "./timerender";
+import type {TopicLink} from "./types";
 import * as user_topics from "./user_topics";
 import * as util from "./util";
 
-function same_day(earlier_msg, later_msg) {
+export type AllVisibilityPolicies = typeof user_topics.all_visibility_policies;
+
+export type MessageContainer = {
+    background_color?: string;
+    date_divider_html: string | undefined;
+    edited_alongside_sender?: boolean;
+    edited_in_left_col?: boolean;
+    edited_status_msg?: boolean;
+    include_recipient: boolean;
+    include_sender: boolean;
+    is_hidden: boolean;
+    mention_classname: string | null;
+    message_edit_notices_in_left_col: boolean;
+    message_edit_notices_alongside_sender: boolean;
+    message_edit_notices_for_status_message: boolean;
+    msg: Message;
+    sender_is_bot: boolean;
+    sender_is_guest: boolean;
+    should_add_guest_indicator_for_sender: boolean;
+    small_avatar_url: string;
+    status_message: string | false;
+    stream_url?: string;
+    subscribed?: boolean;
+    pm_with_url?: string;
+    timestr: string;
+    topic_url?: string;
+    unsubscribed?: boolean;
+    want_date_divider: boolean;
+    last_edit_timestr?: string;
+    moved: boolean;
+    modified: boolean;
+};
+
+export type MessageGroup = {
+    date: string;
+    date_unchanged: boolean;
+    display_recipient: string;
+    message_containers: MessageContainer[];
+    message_group_id: string;
+    bookend_top?: boolean;
+} & (
+    | {
+          is_stream: true;
+          stream_id: number;
+          stream_privacy_icon_color: string;
+          stream_url: string | undefined;
+          recipient_bar_color: string;
+          invite_only: boolean;
+          is_web_public: boolean;
+          visibility_policy: number | false;
+          all_visibility_policies: AllVisibilityPolicies;
+          topic: string;
+          topic_is_resolved: boolean;
+          topic_links: TopicLink[] | undefined;
+          topic_url: string | undefined;
+          match_topic: string | undefined;
+          user_can_resolve_topic: boolean;
+          always_visible_topic_edit: boolean;
+          on_hover_topic_edit: boolean;
+          is_topic_editable: boolean;
+          subscribed: boolean;
+          stream_name: string;
+          just_unsubscribed: boolean;
+          is_subscribed: boolean;
+      }
+    | {
+          is_stream: false;
+          is_private: true;
+          pm_with_url: string | undefined;
+          recipient_users: RecipientRowUser[];
+          display_reply_to_for_tooltip: string;
+      }
+);
+
+type MessageActions = {
```

Comment:
`RenderingPlan` is possibly a better name for this type.

Reply:
Sure, changed to that!